### ISSUE TITLE
BUG: 2nd network not training correctly

### ISSUE THREAD

**Issue by kwhitley** [DanRuta/jsNet#49]
I assume this is a mutation/scope pollution issue, but if I instantiate and train two or more networks (with identical configurations, data sets, etc), one after another, the first trains properly, the second does not appear to train.

Thanks to the speed of your WASM version, I'm able to train models of reasonable accuracy (some are good, some are bad) in only a few seconds... so I plan to let this run for longer, creating many networks, training them, then storing the best scoring networks for use later.

Ideas on what may be causing this?  Are you embedding data into the training/validation data?  Leaving values from previous networks inside the shared window.Module component?

**Comment by DanRuta** [comment]
Hmm, strange that this is happening.

The way it works for WASM is that multiple instances are created over in the C++ code, and the JS instances are assigned an index, which is passed along with any JS->WASM data, including any training/validation/test data, configurations, etc.

The Module is always shared, as there aren't multiple instances of the WASM file loaded.

I've whipped up an example for multiple networks, over in the examples folder. Have a look, maybe it could help, in case the config needs to be different (serve via the server.js file).

Is there anything specific to your configuration which might affect scoping? I don't think I've missed out sending the network index to any emscripten functions, but I'll go through them now, to make sure.

**Comment by DanRuta** [comment]
Couldn't find any missing index parameters (compiler would have screamed at me, anyway). I've tried modifying the multi instance example to do training of subsequent networks in the resolved promise 'then' callback bit, in the case that you set it up that way, but that seemed ok also.

I've tried modifying them to train in parallel also, but that seemed fine also. Is your set-up like either of these?

If the issue lies with the instance indexes, you should be able to see it via net.netInstance (they should all be different for each net instance). If it's not this, then it may be something I can't immediately think of, and would need to analyze a failing example.

**Comment by kwhitley** [comment]
Ok, I think I've isolated it at least... in the callback/promise for the training step, I was using that block to verify results, "score" the network, etc.  It worked totally fine.  I traced through my verification program to see if simply using the network was calling the issue (each verification function called the `net.forward()` with test/validation sample data to score the network, similar to your confusion matrix implementation).  By hardcoding in a prediction rather than using the network itself, the anomaly went away, and future networks were able to train.

To verify simply forwarding the current scoped network was breaking future runs, I removed the verification step and added in a test for loop, running trash data through the network.  Doing this with zero-filled arrays seemed to be fine.  Running actual numbers through it (a randomly filled array) caused the issue.

// Simple version of my implementation (a FC MLP with layers [77, 8, 8, 8, 2])
So the networks are created and "let" scoped within a for loop.  I'm going to show the structure, without the actual config/data because that's not too important:

```js
// outer scope just to create, train, and test networks
for (var i=0; i<10; i++) {
  let net = new Network(networkConfig) // config assumed

  await network.train(trainingdata, trainingConfig).then(() => {
    // here is where I was testing the trained network (via local "net" variable)

    // example purposes only, run 1000 predictions through the network
    for (var j=0; j<1000; j++) {
      net.forward(Array(77).fill(0)) // THIS line works fine for some reason (zero'd input array)
      net.forward(Array(77).fill(0).map(r => Math.random())) // THIS line causes the failures (random input array)
    }

  })
}
```

**Comment by kwhitley** [comment]
Verified that each has a unique instance 👍

**Comment by DanRuta** [comment]
For clarification, where you have `await network.train(train`... , is the `network` variable supposed to be named `net`, or is there a different instance defined, say, before the top for loop?

Also, to answer your other question, which I missed, the training/test data is cleared after use to free memory, so it shouldn't interfere.

> Array(77).fill(0)

TIL

**Comment by kwhitley** [comment]
My bad, that should definitely read:

```js
await net.train(...)
```

Just a typo in the translation.  Really unless there are shared objects/mutability within this library, I see no way for this simple change to have such an effect.  I've gone to the lengths of instantiating *everything*, including the data feeds themselves with locally scoped (to the outer for loop) "let" statements to to be sure I had no mutability in the underlying configs, data, etc.  Seems to all hinge on the forward statement... completely bizarre.

> Array(77).fill(0)

Handy trick I figured out a few weeks ago!  Should note that you can't do this:

```js
Array(77).map(r => 0) // an array of "empty" values can't be iterated for some reason
```

Have to fill it with something before you can map over it, replace values, etc.

**Comment by kwhitley** [comment]
Also, doubt it would make a diff, but here are my network and training configs:

```js
const NETWORK_CONFIG = {
  activation: 'lecuntanh',
  dropout: 1,
  l2: 0.005,
  learningRate: 0.008,
  cost: 'crossEntropy',
  updateFn: 'adagrad',
  weightsConfig: {
    distribution: 'xavieruniform'
  }
}

const TRAINING_CONFIG = {
  epochs: 60,
  shuffle: true,
  maxNorm: true,
  // log: false,
  validation: {
    data: undefined,
    interval: undefined,
    earlyStopping: {
      type: 'threshold',
      threshold: 0.17,
      patience: 10
    }
  }
}
```

validation data and interval are computed and injected just prior to training...

**Comment by DanRuta** [comment]
Configs look fine. It's a strange issue. Logs seem to indicate all instances match up, throughout the function calls. If only diff is actual values, instead of zeroes, the first thing that will be affected will be the sums/activations in layers. I'll try a few things, in the morning.

> Array(77).fill(0)

There's also `[...Array(77)].map(Math.random)`, but last I heard, array destructuring is very slow

**Comment by DanRuta** [comment]
Not sure why I didn't ask this yet, but does this happen on the JS version as well, or just the WASM?

**Comment by DanRuta** [comment]
Ok, I managed to partially recreate differing behaviour, with random data. Messed around with it, and initializing some of the numerical C++ attributes at 0 explicitly instead of leaving them implicitly at 0 got rid of the different behaviour. I'm new to C++.

Try out v3.3.5, let me know if this fixed it.

**Comment by kwhitley** [comment]
Seems to have fixed it!  Thanks so much for the quick dig/fix - seems crazy that something so minor would affect it like that... 

Time to create all the networks!  Muahahaha

**Comment by kwhitley** [comment]
Totally working like a champ now!  Thanks so much!!!

![image](https://user-images.githubusercontent.com/865416/38099295-da7b258c-333f-11e8-8793-646a9e1b688e.png)

**Comment by kwhitley** [comment]
Nah, I use the term "generation" loosely here - although I have definitely played with neuroevolution using the neataptic library.

So what I'm actually working on - and the randomish nature of the data makes testing very difficult, is creating short term predictors for the major crypto markets.  The point/plan is two-fold:

1. Prove there is some predictable nature to the crypto markets - many currently believe that nearly all financial markets are pure random-walk/martingale process, meaning you have perfectly even chance of guessing correctly (no better than flipping a coin).  A buddy flat out told me it was impossible to predict and that I'd fail.  So if nothing else, it's to prove him wrong.  So far I've used your library (the speed helps the iteration cycle of experimentation sooooo much) to achieve a binary classifier on ETH that can determine with ~57% accuracy whether the price will go up or down in the next 2 hours (based on a total data set of 3000 hours).  That's certainly beyond the range of statistical error.

2. Ultimately create day-trader bots trained to buy/sell via RL or neuro-evolution.  I've done this already to varying degrees of success, but trying to teach it how to trade at the same time as determining patterns under the financial data was proving problematic.  If I can separate the tasks, using traditional networks as prediction indicators, and feed that to the evolutionary algos, it may be more effective.

3. At the moment I'm recording batches of the best MLP predictors, with the goal of feeding their predictions as inputs into a secondary network (again, with the original expected values)... like a "voting party".  Then see if I can train the 2nd network to achieve better overall results based on this.  Fun experiment if nothing else!

Sidenote: There may be a memory leak in there somewhere (could be my code too)... seems to happen all the time right around generation/network ~#720
![image](https://user-images.githubusercontent.com/865416/38114634-be95819e-336e-11e8-8e33-c5add8318152.png)

**Comment by kwhitley** [comment]
Oh yeah, and if Imanage to crack this open using your lib... expect some kickbacks :)

**Comment by DanRuta** [comment]
Ah, interesting project. My housemate was working on something like that for his uni final year project. I'd have thought an LSTM model might be good, given time series data (eg chart patterns, like 'head and shoulders', 'cup and handle', etc), but I've never actually tried this myself. Even twitter sentiment analysis might help, given the impact of news on the market. I'd be very interested to know if it works out, do let me know! 😊

I'll have to play around with neaptic a bit, I've definitely been thinking of adding neuro-evolution to jsNet.


It's quite likely I missed out some C++ memory cleaning, I'll go through it this weekend, I've experienced it myself, also. In the meantime, if you have the `collectErrors` config turned to true (default false), you can toggle it off, as that definitely aggregates data, by design.

**Comment by kwhitley** [comment]
LSTM would def be good for it, but most of the JS libs that support that (e.g. synaptic, neataptic) are not optimized... so running any large dataset through it is incredibly slow.  The ones that *are* accelerated (e.g. Google's GPU-accelerated deeplearn.js, which I've written a high level wrapper for) have typically been extremely feature-bare, and only basic MLP support.  In light of that, I'm using an old fashioned moving-window approach with an MLP.  It's certainly less advanced than a proper LSTM implementation, but seems reasonably effective in the meantime

Neuro-evolution is *incredible* in terms of the speed at which it can evolve and learn a dataset (even non-hardware accelerated, but the only library with a great interface (neataptic), do not seem to offer good built-in validation tools for the evolutionary paths.  That's extremely important because the neuro-evolution can overfit SO quickly.

Your lib was so attractive because the incredible amount of built in features (dropout, l2, l1, dataset splitting, validation, stopping, etc, etc) with a great API/interface AND hardware acceleration.  I'm able to pipe in 10x the data with networks 10-20x as large (e.g. current input layer of ~220) and still get relatively quick training (a few seconds to stop point).  If you implemented LSTM + neuro, your library would stand a chance of being *the* main ML library for JS development.  I'm super excited for this!

As to my project... I just had a huge breakthrough last night... turns out LTC can be strongly predicted based on the current movement of LTC+ETH+BTC (the others are more difficult, with BTC being the most difficult, implying that it "leads" the market, and others move based on its movement).  I'm able to reliably predict the future up/down movement of LTC to 75%+ accuracy (at a softmax threshold of only >0.5).  That's kinda crazy given these markets are alleged to be so random...

Definitely intend to dive into news/sentiment, but as a later step!

**Comment by DanRuta** [comment]
I'm pretty interested in learning more about LSTM, and I've already used genetic algorithms while studying at uni, so time permitting, expect to see more. I'm also deeply intrigued by GANs, but that's a ways off, still.

I'm sending off applications now to do a Masters in AI stuff next academic year, and plan to add new stuff to jsNet, as I learn more. For now, the focus is on GPU acceleration, though. 

That LTC performance is insane, actually. For the others, maybe it would be easier to consider value against BTC rather than FIAT.

**Comment by DanRuta** [comment]
I was thinking, maybe that memory limit was reached from there being too many active instances (each taking up a whole bunch of memory for their data). I've added a `net.delete()` function, which should clear the memory of its net's instance, in the WASM code, after which `delete net` can be used to get rid of the js code, too.

If it's possible to get rid of old nets in your set-up, I think this should do it.

**Comment by kwhitley** [comment]
I assumed it was something along those lines, as I can generate fewer bigger networks than smaller ones (before failure).  I'm declaring and training locally scoped (let) nets inside a for loop, so they should be garbage collected automatically as each pass will lose the previous reference.  Could be totally wrong though.

**Comment by DanRuta** [comment]
That's true for the JavaScript code. The memory taken up by the stuff in the WebAssembly code, however, is stored in the typed "heap" arrays (Can be seen via, for example `Module.HEAP32`). This is not scoped to user created JavaScript code, I don't believe, so it persists. 

By calling the `delete()` function, you'd be calling the appropriate functions in C++ that would clear the data (weights, deltas, biases, etc) from these heap arrays, which represent the emulated memory locations. 

It would be strange, if there would be memory leaks, as no new data is allocated without being freed up, once a network is initialized, and run once. Data (weights, deltas, etc) just gets mutated. I made this little library (https://github.com/DanRuta/JS-Memory-Map) to visualize memory usage, and, using a tiny network, it didn't seem to increase.
